File: output/PrimesFFI.php
<?php
$exports['runPrimesFFI'] = function($limit) {
    $n = (int)$limit;
    
    $range = function($start) {
        return function($end) use ($start) {
            $go = function($curr, $acc) use (&$go, $start) {
                if ($curr < $start) return $acc;
                return (object)["type" => "Cons", "value0" => $curr, "value1" => $acc];
            };
            return $go($end, (object)["type" => "Nil"]);
        };
    };

    $filter = function($p) {
        return function($lst) use ($p) {
            $go = function($list, $acc) use (&$go, $p) {
                if ($list->type === "Nil") {
                    $rev = function($l, $a) use (&$rev) {
                        if ($l->type === "Nil") return $a;
                        return $rev($l->value1, (object)["type" => "Cons", "value0" => $l->value0, "value1" => $a]);
                    };
                    return $rev($acc, (object)["type" => "Nil"]);
                }
                $x = $list->value0;
                $xs = $list->value1;
                if ($p($x)) {
                    return $go($xs, (object)["type" => "Cons", "value0" => $x, "value1" => $acc]);
                } else {
                    return $go($xs, $acc);
                }
            };
            return $go($lst, (object)["type" => "Nil"]);
        };
    };

    $sieve = function($lst) use (&$sieve, $filter) {
        if ($lst->type === "Nil") return (object)["type" => "Nil"];
        $p = $lst->value0;
        $xs = $lst->value1;
        $filtered = $filter(function($x) use ($p) { return $x % $p !== 0; })($xs);
        return (object)[
            "type" => "Cons",
            "value0" => $p,
            "value1" => $sieve($filtered)
        ];
    };

    $sumList = function($lst) {
        $go = function($list, $acc) use (&$go) {
            if ($list->type === "Nil") return $acc;
            return $go($list->value1, $acc + $list->value0);
        };
        return $go($lst, 0);
    };

    $rng = $range(2)($n);
    $s = $sieve($rng);
    return $sumList($s);
};
return $exports;

File: scratch/compact-guards-integration-20260916/after/src/Test/PrimesFFI.erl
-module(test_primesFFI@foreign).
-export([runPrimesFFI/1]).
range(Start, Curr, Acc) when Curr < Start -> Acc;
range(Start, Curr, Acc) -> range(Start, Curr - 1, {cons, Curr, Acc}).
reverse({nil}, Acc) -> Acc;
reverse({cons, X, Xs}, Acc) -> reverse(Xs, {cons, X, Acc}).
filter(_P, {nil}, Acc) -> reverse(Acc, {nil});
filter(P, {cons, X, Xs}, Acc) ->
  case P(X) of
    true -> filter(P, Xs, {cons, X, Acc});
    false -> filter(P, Xs, Acc)
  end.
sieve({nil}) -> {nil};
sieve({cons, P, Xs}) ->
  {cons, P, sieve(filter(fun(X) -> X rem P =/= 0 end, Xs, {nil}))}.
sum_list({nil}, Acc) -> Acc;
sum_list({cons, X, Xs}, Acc) -> sum_list(Xs, Acc + X).
runPrimesFFI(N) -> sum_list(sieve(range(2, N, {nil})), 0).
